File: src/main/java/com/aaronhowser1/pitchperfect/enchantment/BwaaapEnchantment.java
package com.aaronhowser1.pitchperfect.enchantment;

import com.aaronhowser1.pitchperfect.config.CommonConfigs;
import com.aaronhowser1.pitchperfect.utils.ServerUtils;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.enchantment.Enchantment;
import net.minecraft.world.item.enchantment.EnchantmentCategory;
import net.minecraft.world.phys.Vec3;

import java.util.List;

public class BwaaapEnchantment extends Enchantment {

    public BwaaapEnchantment(Enchantment.Rarity pRarity, EnchantmentCategory pCategory, EquipmentSlot... pApplicableSlots) {
        super(pRarity, pCategory, pApplicableSlots);
    }

    @Override
    public int getMinCost(int pLevel) {
        return 5;
    }

    @Override
    public int getMaxCost(int pLevel) {
        return 25;
    }

    public static List<LivingEntity> getTargets(LivingEntity user) {
        int range = CommonConfigs.BWAAAP_RANGE.get();

        return ServerUtils.getNearbyLivingEntities(user, range);
    }

    public static void knockBack(LivingEntity user) {

        for (LivingEntity target : getTargets(user)) {
            int range = CommonConfigs.BWAAAP_RANGE.get();
            float strength = CommonConfigs.BWAAAP_STRENGTH.get();

            Vec3 targetMotion = target.getDeltaMovement();
            Vec3 userToTargetVector = new Vec3(
                    target.getX()-user.getX(),
                    target.getY()-user.getY(),
                    target.getZ()-user.getZ()
            );
            double distanceToTarget = userToTargetVector.length();
            double targetPercentageFromRange = Math.abs(distanceToTarget/range - 1);

            float yMult = user.isCrouching() ? 2F : 1F;

            target.setDeltaMovement(
                    targetMotion.add(
                            userToTargetVector.multiply(
                                    targetPercentageFromRange*strength,
                                    targetPercentageFromRange*strength*yMult,
                                    targetPercentageFromRange*strength
                            )
                    )
            );
        }
    }


}

File: src/main/kotlin/com/aaronhowser1/pitchperfect/enchantment/BwaaapEnchantment.kt
package com.aaronhowser1.pitchperfect.enchantment

import com.aaronhowser1.pitchperfect.config.ServerConfig
import com.aaronhowser1.pitchperfect.item.InstrumentItem
import com.aaronhowser1.pitchperfect.utils.ServerUtils.distanceBetweenPoints
import com.aaronhowser1.pitchperfect.utils.ServerUtils.entityToVec3
import com.aaronhowser1.pitchperfect.utils.ServerUtils.getNearbyLivingEntities
import com.aaronhowser1.pitchperfect.utils.ServerUtils.vecBetweenPoints
import net.minecraft.world.entity.EquipmentSlot
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.enchantment.Enchantment
import kotlin.math.abs
import kotlin.math.roundToInt

object BwaaapEnchantment : Enchantment(
    Rarity.COMMON,
    ModEnchantments.INSTRUMENT,
    arrayOf(
        EquipmentSlot.MAINHAND,
        EquipmentSlot.OFFHAND
    )
) {

    override fun getMinCost(pLevel: Int): Int = 5
    override fun getMaxCost(pLevel: Int): Int = 25

    fun triggerBwaaap(user: LivingEntity, instrument: InstrumentItem) {
        BwaaapTrigger(user, instrument)
    }

    private class BwaaapTrigger(val user: LivingEntity, instrument: InstrumentItem) {
        private val targets: List<LivingEntity> = getTargets()
        private val cooldown: Int = getCooldown()

        private val range: Float = ServerConfig.BWAAAP_RANGE.get()


        init {
            knockBack()
            if (user is Player) user.cooldowns.addCooldown(instrument, cooldown)
        }

        private fun getTargets(): List<LivingEntity> {
            return getNearbyLivingEntities(user, range)
        }

        private fun getCooldown(): Int {
            var cooldown = 0f

            for (target in targets) {
                val distanceToTarget = distanceBetweenPoints(
                    entityToVec3(user),
                    entityToVec3(target)
                )
                if (distanceToTarget > range) continue

                val targetPercentageFromRange = abs(distanceToTarget / range - 1)
                cooldown += targetPercentageFromRange.toFloat()
            }

            cooldown *= ServerConfig.BWAAAP_COOLDOWN_MULT.get()
            return cooldown.roundToInt()
        }

        private fun knockBack() {
            val strength: Float = ServerConfig.BWAAAP_STRENGTH.get()

            for (target in targets) {
                val targetMotion = target.deltaMovement

                val userToTargetVector = vecBetweenPoints(
                    entityToVec3(user),
                    entityToVec3(target)
                )

                val distanceToTarget = userToTargetVector.length()
                if (distanceToTarget > range) continue

                val targetPercentageFromRange = abs(distanceToTarget / range - 1)
                val yMult = if (user.isCrouching) 2f else 1f

                target.deltaMovement = targetMotion.add(
                    userToTargetVector.multiply(
                        targetPercentageFromRange * strength,
                        targetPercentageFromRange * strength * yMult,
                        targetPercentageFromRange * strength
                    )
                )
            }
        }

    }

}